src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,192 @@
+                if (context.DataSource.ResolvedProtocol is DataSource.Protocol.TCP)
+                {
+                    context.ConnectionStream = await HandleTcpRequest(context, isAsync, ct).ConfigureAwait(false);
+                }
+            }
+            catch (Exception e)
+            {
+                context.Error = e;
+            }
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        private ValueTask<Stream> HandleNamedPipeRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private ValueTask<Stream> HandleSharedMemoryRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private async ValueTask<Stream> HandleTcpRequest(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            var ipAddresses = isAsync
+                ? await Dns.GetHostAddressesAsync(context.DataSource.ServerName, ct).ConfigureAwait(false)
+                : Dns.GetHostAddresses(context.DataSource.ServerName);
+            if (ipAddresses.Length == 0)
+            {
+                throw new Exception("Hostname did not resolve");
+            }
+
+            // If there is an IP version preference, apply it
+            switch (context.IpAddressPreference)
+            {
+                case SqlConnectionIPAddressPreference.IPv4First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV4);
+                    break;
+
+                case SqlConnectionIPAddressPreference.IPv6First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV6);
+                    break;
+
+                case SqlConnectionIPAddressPreference.UsePlatformDefault:
+                default:
+                    // Not sorting necessary
+                    break;
+            }
+
+            // Attempt to connect to one of the matching IP addresses
+            // @TODO: Handle opening in parallel
+            Socket socket = null;
+            var socketOpenExceptions = new List<Exception>();
+
+            var ipEndpoint = new IPEndPoint(IPAddress.None, context.DataSource.Port); // Allocate once
+            foreach (var ipAddress in ipAddresses)
+            {
+                ipEndpoint.Address = ipAddress;
+                try
+                {
+                    socket = await OpenSocket(ipEndpoint, isAsync, ct).ConfigureAwait(false);
+                    break;
+                }
+                catch(Exception e)
+                {
+                    socketOpenExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                throw socketOpenExceptions.OfType<SocketException>().FirstOrDefault()
```

Comment:
Nit: very small reservation about adding a dependency on LINQ at this stage.

Reply:
I feel like it's safe since it's not a fast path. But then again, this will be used in parallel for multi-subnet failover where failures will be expected. Maybe this is a fast path in that case? I just like the brevity of the code, but if the performance really is garbage, I can be convinced to rewrite as a loop.

Reply:
I'm not thinking of performance - as far as I know, that's broadly fine, and I agree that it's not on a hot path, and that variations connection/network latency will probably outweigh the LINQ CPU overhead. There's potentially a point to be made around allocations, but I don't expect those to be particularly large.

My motivation's purely around avoiding unnecessary dependencies on packages. If everything else in the codebase compiles without a reference to System.Linq then I think it'd be better to use a loop (or a variable to track the first SocketException as it's raised) - the change is simple enough that it's not worth dragging in another package for. If SqlClient's already using LINQ, I've got no objections.

Reply:
I +1 this comment for LINQ removal due to @JRahnama 's comment. 

My personal Point of view: 
LINQ has great usability, but sometimes the usability comes with a perf overhead. Knowing which is a hot path and which is not, is something that the domain experts only understand. (in this case pretty much all the reviewers of the code). I am hoping to stay away from "That part of driver uses LINQ, so can this" in future. Typically when code becomes older, the ways of doing things in a code become a standard, rather than deeper reasoning. Because of this I personally would recommend against usage of LINQ.

Reply:
Affirm, will do.

(but I still love linq 🥹)